### PR TITLE
Adds meta pixel settings section

### LINKED ISSUE #1554: Add field for FB Pixel ID
If a site is AMP Plus, show field for Facebook Pixel ID in Engagement > Social. Using this ID, place the following tag just before the closing head tag. (Note there are **two** locations for this ID to be inserted within the tag)

```
<!-- Meta Pixel Code -->
<script>
!function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod?
n.callMethod.apply(n,arguments):n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window, document,'script',
'https://connect.facebook.net/en_US/fbevents.js');
fbq('init', 'PIXELID');
fbq('track', 'PageView');
</script>
<noscript><img height="1" width="1" style="display:none"
src="https://www.facebook.com/tr?id=PIXELID&ev=PageView&noscript=1"
/></noscript>
<!-- End Meta Pixel Code -->
```

Background: After internal discussion of the difficulties for publishers trying to place an FB pixel, it was determined that the best solution would be to provide an ID field and place the pixel on their behalf rather than recommending a third party plugin to do so. Find full discussion in P2 here: pamTN9-3ZQ-p2

### PR DESCRIPTION
**Pull request by leogermani**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #1554 .

Relevant discussion: https://wp.me/pamTN9-3ZQ

Adds an option to the Engagement > Social settings page to allow users to configure Meta (Facebook) pixel.

### How to test the changes in this Pull Request:

1. Go to Newspack > Engagement > Social
2. Activate "Meta Pixel"
3. Add a pixel ID
4. Save
5. Test invalid formats
6. Test enabling and disabling the option
7. Check the copy and the link
8. Visit the site and confirm the pixel was added to the page. You can use a browser extension to help you with that if you want: https://href.li/?https://chrome.google.com/webstore/detail/facebook-pixel-helper/fdgfkebogiimcoedlicjlajpkdmockpc?hl=en
9. Visit the site with AMP enabled and confirm the pixel is also there
10. I've only tested this with random IDs. It would be awesome if someone could test it with a real pixel ID in a real site

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by adekbadek — COMMENTED** [review]
Please add a PR description here – it will help to debug any related issues later, as well as serve as a description of the feature to any non-developers.

**Comment by leogermani** [comment]
> Please add a PR description here – it will help to debug any related issues later, as well as serve as a description of the feature to any non-developers.

It's still in progress :)

**Comment by adekbadek** [comment]
Ah, all right! If you submit a PR, our GH team gets a ping to review it. Until the PR is in progress, please submit it as a draft PR :)

**Comment by leogermani** [comment]
> 

ok, noted. I thought you would only be notified if I added the Needs Review label

**Comment by adekbadek** [comment]
That's another level of notifications – our team is marked as [code owners](https://github.com/Automattic/newspack-plugin/blob/b6014167449c252c982b123e0d6e42ad167e4fde/.github/CODEOWNERS#L0-L1), so we all get a GH ping when a PR is submitted. Then when the label is added, that pings Slack. This is all to ensure PRs are reviewed as quickly as possible.

File: includes/class-meta-pixel.php
Comment (adekbadek) [inline includes/class-meta-pixel.php:157]: The parameters of the URL in the `src` should be variables, or omitted. There are a few which are dynamic – as e.g. `event_url` here with test site URL value ;)

File: includes/wizards/class-engagement-wizard.php
Comment (adekbadek) [inline includes/wizards/class-engagement-wizard.php:151]: When the option is not yet saved, an API request will be sent on clicking the toggle. This request will fail, since the `pixel_id` is invalid (it's `""` by default). This makes the toggle un-toggleable. 

image

**Review by adekbadek — CHANGES_REQUESTED** [review]

File: includes/class-meta-pixel.php
Comment (leogermani) [reply includes/class-meta-pixel.php:157]: oh sorry I missed that.

The JS script version of the snippet does not include any of these variables. I'm gonna remove them from here for now so we can validate if the whole thing works and we can improve it later

Comment (leogermani) [reply includes/class-meta-pixel.php]: actually, decided to add the variables there

File: includes/class-meta-pixel.php
Comment (miguelpeixe) [inline includes/class-meta-pixel.php:163]: Are these values standard? Looks like the parameters being used for the img snippet are not being applied to the script. Won't this cause inconsistent tracking?

It would be nice to have them filterable in an isolated private method like `get_snippet_parameters()` to be used in both img and script snippets.

File: includes/class-meta-pixel.php
Comment (miguelpeixe) [inline includes/class-meta-pixel.php:198]:
```suggestion
	 * Prints snippets in the header
```

File: includes/class-meta-pixel.php
Comment (miguelpeixe) [inline includes/class-meta-pixel.php:224]: AMP enabled does not mean the browser doesn't support javascript. A `! self::is_amp()` check is more suitable for this.

**Review by miguelpeixe — CHANGES_REQUESTED** [review]

File: includes/class-meta-pixel.php
Comment (leogermani) [reply includes/class-meta-pixel.php:224]: This is not a check for AMP enabled.

The `noscript` tag is there for non-AMP site. If visiting a non AMP site without JS, the IMG tag will be used for tracking.

If AMP is enabled, the AMP plugin will capture this and replace the IMG with a `<amp-pixel>` tag. So this covers both cases and we don't need to manually add the amp tag. We leave it to the AMP plugin

File: includes/class-meta-pixel.php
Comment (leogermani) [reply includes/class-meta-pixel.php:163]: All examples I've seen of the JS snippet do not include these vars...

Comment (leogermani) [reply includes/class-meta-pixel.php:163]: Added the custom properties to the JS snippet following what I see here:

https://developers.facebook.com/docs/meta-pixel/implementation/conversion-tracking#custom-properties

Or we could also remove them from both places... The request in the [original issue](https://github.com/Automattic/newspack-plugin/issues/1554) did not ask for these properties.

I think this is something I brought looking at one of the plugins that add the pixel.

Comment (miguelpeixe) [reply includes/class-meta-pixel.php:163]: I believe it's good practice to implement a default set of parameters and make it filterable! This is working well 👍

File: includes/class-meta-pixel.php
Comment (miguelpeixe) [inline includes/class-meta-pixel.php:165]:
```suggestion
		if ( is_singular() ) {
```

Small tweak include these parameters to `page` types

**Review by miguelpeixe — CHANGES_REQUESTED** [review]

**Review by miguelpeixe — APPROVED** [review]
🙌

**Pull request merged**
